src/neml2/dispatchers/SimpleMPIScheduler.cxx

```diff
@@ -0,0 +1,165 @@
+
+  options.set<std::size_t>("batch_size");
+  options.set("batch_size").doc() = "Batch size";
+
+  options.set<std::size_t>("capacity");
+  options.set("capacity").doc() =
+      "Maximum number of work items that can be dispatched, default to batch_size";
+
+  return options;
+}
+
+SimpleMPIScheduler::SimpleMPIScheduler(const OptionSet & options)
+  : WorkScheduler(options),
+    _available_devices(options.get<std::vector<Device>>("devices")),
+    _batch_size(options.get<std::size_t>("batch_size")),
+    _capacity(options.get("capacity").user_specified() ? options.get<std::size_t>("capacity")
+                                                       : _batch_size),
+    _comm(TIMPI::Communicator(MPI_COMM_WORLD))
+{
+}
+
+void
+SimpleMPIScheduler::setup()
+{
+  WorkScheduler::setup();
+
+  const auto & device_list = input_options().get<std::vector<Device>>("devices");
+
+  // First pass:
+  // - Prohibit any CPU
+  // - Check if any CUDA device is present
+  bool cuda = false;
+  for (const auto & device : device_list)
+  {
+    neml_assert(!device.is_cpu(), "CPU device is not allowed in SimpleMPIScheduler");
+    if (device.is_cuda())
+      cuda = true;
+    else
+      neml_assert(false, "Unsupported device type: ", device);
+  }
+
+  // Second pass:
+  // - If multiple CUDA devices are present, make sure each CUDA device has a concrete
+  //   (nonnegative), unique device ID
+  bool has_multiple_cuda_devices = device_list.size() > 1 && cuda;
+  if (has_multiple_cuda_devices)
+  {
+    std::set<DeviceIndex> cuda_device_ids;
+    for (const auto & device : device_list)
+    {
+      if (device.is_cpu())
+        continue;
+      auto device_id = device.index();
+      neml_assert(device_id >= 0, "Device ID must be nonnegative");
+      neml_assert(cuda_device_ids.find(device_id) == cuda_device_ids.end(),
+                  "Device ID must be unique. Found duplicate: ",
+                  device_id);
+      cuda_device_ids.insert(device_id);
+    }
+  }
+
+  determine_my_device();
+}
+
+void
+SimpleMPIScheduler::determine_my_device()
+{
+  // Get the hostname via hwloc and hash it into an MPI_INT
+  hwloc_topology_t topology;
+  hwloc_obj_t machine;
+
+  hwloc_topology_init(&topology); // initialization
+  hwloc_topology_load(topology);  // actual detection
+  machine = hwloc_get_root_obj(topology);
+  std::hash<std::string> hasher;
+  std::string hostname(hwloc_obj_get_info_by_name(machine, "HostName"));
+  int id = static_cast<int>(hasher(hostname) % std::numeric_limits<int>::max());
+  hwloc_topology_destroy(topology);
+
+  // Make a new communicator based on this hashed hostname
+  TIMPI::Communicator new_comm;
+  _comm.split(id, _comm.rank(), new_comm);
+  // Assign our device index based on the new communicator
+  _device_index = new_comm.rank();
+  neml_assert(_device_index < _available_devices.size(),
+              "MPI split by host would require too many devices");
```

Comment:
I understand `_device_index == _available_devices.size() - 1`.
I also understand why `_device_index > _available_devices.size() - 1` should error.

How about the other case `_device_index < _available_devices.size() - 1`? Does it mean one of the devices will remain idle?

Reply:
I fixed the docstring, i'm fine with a device being idle.  Though maybe we should throw a warning?